File: android/src/main/java/com/sip/flutter/sip_sdk_flutter/sip/SIPManage.java
package com.sip.flutter.sip_sdk_flutter.sip;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import com.sip.flutter.sip_sdk_flutter.SipSdkFlutterPlugin;
import com.sip.flutter.sip_sdk_flutter.utils.camera.CameraHandle;
import com.sip.flutter.sip_sdk_flutter.utils.camera.CameraStateChangeCallback;
import com.sip.sdk.SIPSDK;
import com.sip.sdk.entity.SIPSDKCallParam;
import com.sip.sdk.entity.SIPSDKCallStatusParam;
import com.sip.sdk.entity.SIPSDKConfig;
import com.sip.sdk.entity.SIPSDKDtmfInfoParam;
import com.sip.sdk.entity.SIPSDKFindIncomingParam;
import com.sip.sdk.entity.SIPSDKFTCompleteParam;
import com.sip.sdk.entity.SIPSDKFTOfferParam;
import com.sip.sdk.entity.SIPSDKFTProgress;
import com.sip.sdk.entity.SIPSDKFTRequestInfo;
import com.sip.sdk.entity.SIPSDKMediaConfig;
import com.sip.sdk.entity.SIPSDKMessageParam;
import com.sip.sdk.i.SIPSDKFTListener;
import com.sip.sdk.i.SIPSDKListener;

import java.util.HashMap;
import java.util.Map;

public class SIPManage implements SIPSDKListener.InitCompletedListener,
        SIPSDKListener.RegistryStateListener,
        SIPSDKListener.DtmfInfoListener,
        SIPSDKListener.MessageListener,
        SIPSDKListener.MessageStateListener,
        SIPSDKListener.IncomingCallListener,
        SIPSDKListener.FindIncomingListener,
        SIPSDKListener.CallStateListener,
        SIPSDKListener.ExpireWarningCallbackListener,
        SIPSDKListener.ActivityCheckCallbackListener,
        SIPSDKFTListener.FTRequestListener,
        SIPSDKFTListener.FTRequestResultListener,
        SIPSDKFTListener.FTOfferListener,
        SIPSDKFTListener.FTProgressListener,
        SIPSDKFTListener.FTCompleteListener,
        CameraStateChangeCallback {
    private final Handler handler = new Handler(Looper.getMainLooper());

    // FT 进度节流状态：进度回调按块高频触发，转发前按 150ms 合并（见 onProgress）。
    private long lastFTProgressMs = 0;
    private long lastFTProgressId = -1L;

    private static class Instance {
        private static final SIPManage instance = new SIPManage();
    }

    public static SIPManage instance() {
        return Instance.instance;
    }

    public void init(Context context,
                     String baseUrl,
                     String clientId,
                     String clientSecret,
                     SIPSDKConfig config,
                     SIPSDKMediaConfig mediaConfig) {
        //注册摄像头状态监听
        CameraHandle.instance().addStateChangeCallback(this);
        //注册SDK回调
        SIPSDK.addListener(this);
        //注册文件传输回调
        SIPSDK.addFTListener(this);
        //初始化SDK
        SIPSDK.init(context, baseUrl, clientId, clientSecret, config, mediaConfig);
    }

    public void initToken(
            String token,
            String clientId,
            String clientSecret,
            SIPSDKConfig config,
            SIPSDKMediaConfig mediaConfig) {
        //注册摄像头状态监听
        CameraHandle.instance().addStateChangeCallback(this);
        //注册SDK回调
        SIPSDK.addListener(this);
        //注册文件传输回调
        SIPSDK.addFTListener(this);
        //初始化SDK
        SIPSDK.initToken(token, clientId, clientSecret, config, mediaConfig);
    }

    @Override
    public void onStateChange(boolean runing) {
        Map<String, Object> map = new HashMap<>();
        map.put("state", runing);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onCameraStateChange", map);
        });
    }

    @Override
    public void onExpireWarning(long expireTime, long currentTime) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("expireTime", expireTime);
        payload.put("currentTime", currentTime);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onExpireWarning", payload);
        });
    }

    @Override
    public void onRegistryState(int code) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("state", code);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onRegistrarState", payload);
        });
    }

    @Override
    public void onInitCompleted(int state, String msg) {
        Map<String, Object> map = new HashMap<>();
        map.put("state", state);
        map.put("message", msg);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onInitCompleted", map);
        });
    }

    @Override
    public void onIncomingCall(SIPSDKCallParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("callType", param.callType);
        payload.put("username", param.username);
        payload.put("remoteIp", param.remoteIp);
        payload.put("headers", param.headers);
        payload.put("callUuid", String.valueOf(param.callUuid));
        payload.put("transmitVideo", param.transmitVideo);
        payload.put("transmitSound", param.transmitSound);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onIncomingCall", payload);
        });
    }

    @Override
    public int onFindIncoming(SIPSDKFindIncomingParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("currentType", param.currentType);
        payload.put("transportType", param.transportType);
        payload.put("transportName", param.transportName);
        payload.put("toDomain", param.toDomain);
        payload.put("toUsername", param.toUsername);
        payload.put("fromDomain", param.fromDomain);
        payload.put("fromUsername", param.fromUsername);
        payload.put("requestDomain", param.requestDomain);
        payload.put("requestUsername", param.requestUsername);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onFindIncoming", payload);
        });
        return param.currentType;
    }

    @Override
    public void onCallState(SIPSDKCallStatusParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("state", param.state);
        payload.put("callUuid", String.valueOf(param.callUuid));
        payload.put("lastStatus", param.lastStatus);
        payload.put("lastStatusText", param.lastStatusText);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onCallState", payload);
        });
    }

    @Override
    public void onDtmfInfo(SIPSDKDtmfInfoParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("callUuid", String.valueOf(param.callUuid));
        payload.put("dtmfInfoType", param.dtmfInfoType);
        payload.put("contentType", param.contentType);
        payload.put("content", param.content);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onDtmfInfo", payload);
        });
    }

    @Override
    public void onMessage(SIPSDKMessageParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("messageType", param.messageType);
        payload.put("username", param.username);
        payload.put("remoteIp", param.remoteIp);
        payload.put("content", param.content);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onMessage", payload);
        });
    }

    @Override
    public void onMessageState(int state, SIPSDKMessageParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("state", state);
        Map<String, Object> message = new HashMap<>();
        message.put("messageType", param.messageType);
        message.put("username", param.username);
        message.put("remoteIp", param.remoteIp);
        message.put("content", param.content);
        payload.put("message", message);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onMessageState", payload);
        });
    }

    @Override
    public void onActivityCheck() {
        Map<String, Object> payload = new HashMap<>();
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onActivityCheck", payload);
        });
    }

    // ---- 文件传输（FT） ----

    private Map<String, Object> ftFileMetaMap(com.sip.sdk.entity.SIPSDKFTFileMeta file) {
        Map<String, Object> fileMap = new HashMap<>();
        if (file == null) {
            fileMap.put("name", "");
            fileMap.put("size", 0L);
            fileMap.put("extra", "");
            return fileMap;
        }
        fileMap.put("name", file.name);
        fileMap.put("size", file.size);
        fileMap.put("extra", file.extra);
        return fileMap;
    }

    @Override
    public void onOffer(SIPSDKFTOfferParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("ftId", param.ftId);
        payload.put("accUuid", param.accUuid);
        payload.put("username", param.username);
        payload.put("remoteIp", param.remoteIp);
        payload.put("file", ftFileMetaMap(param.file));
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onFTOffer", payload);
        });
    }

    @Override
    public void onRequest(SIPSDKFTRequestInfo info) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("reqId", info.reqId);
        payload.put("accUuid", info.accUuid);
        payload.put("username", info.username);
        payload.put("remoteIp", info.remoteIp);
        payload.put("file", ftFileMetaMap(info.file));
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onFTRequest", payload);
        });
    }

    @Override
    public void onRequestResult(long reqId, boolean ok, String reason) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("reqId", reqId);
        payload.put("ok", ok);
        payload.put("reason", reason);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onFTRequestResult", payload);
        });
    }

    @Override
    public void onProgress(SIPSDKFTProgress progress) {
        // 进度回调按块触发（1.5MB≈数百~上千次），且 Dart 侧当前只把它当一次
        // 「攒够 minBytes 就边传边播」的触发信号，不展示进度条；若全量经主线程
        // handler.post + MethodChannel 转发，会往主线程灌几百次跨 isolate 往返，
        // 明显拉高 CPU 并卡 UI。这里按 150ms 合并，一次传输最多转发约十次。
        // （新会话首个事件不受节流限制，保证起始进度能及时到达。）
        long now = System.currentTimeMillis();
        if (now - lastFTProgressMs < 150 && progress.ftId == lastFTProgressId) {
            return;
        }
        lastFTProgressMs = now;
        lastFTProgressId = progress.ftId;
        Map<String, Object> payload = new HashMap<>();
        payload.put("ftId", progress.ftId);
        payload.put("state", progress.state);
        payload.put("bytesTotal", progress.bytesTotal);
        payload.put("bytesDone", progress.bytesDone);
        payload.put("percent", progress.percent);
        payload.put("activeSessions", progress.activeSessions);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onFTProgress", payload);
        });
    }

    @Override
    public void onComplete(SIPSDKFTCompleteParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("ftId", param.ftId);
        payload.put("role", param.role);
        payload.put("error", param.error);
        payload.put("bytesTransferred", param.bytesTransferred);
        payload.put("elapsedMs", param.elapsedMs);
        payload.put("fileName", param.fileName);
        payload.put("savePath", param.savePath);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onFTComplete", payload);
        });
    }
}

File: ios/Classes/SIP/SIPManage.swift
//
//  SIPManage.swift
//  SIPSDKExample
//
//  Created by 杨涛 on 2025/5/18.
//

import AVFoundation
import SIPFramework

public extension Notification.Name {
    static let SIP_CALL_STATE_CHANGE = Notification.Name("SIP_CALL_STATE_CHANGE")
}

public class SIPManage {
    // 定义符合 C 调用约定的回调函数
    static let onLogCallback: OnLogCallback = { level, data, _ in
        let levelText: String
        switch level {
        case 0: levelText = "DEBUG"
        case 1: levelText = "INFO"
        case 2: levelText = "WARN"
        case 3: levelText = "ERROR"
        default: levelText = "UNKNOWN"
        }

        if let data = data {
            let message = String(cString: data)
            print("[\(levelText)] \(message)")
        } else {
            print("[\(levelText)] <no message>")
        }
    }

    static let onInitCompleted: OnInitCompleted = { state, msg in
        var message: String?
        if let msg = msg {
            message = String(cString: msg)
            print("SDK Init completed. State: \(state), Message: \(message ?? "nil")")
        } else {
            print("SDK Init completed. State: \(state), No message.")
        }
        let payload: [String: Any] = [
            "state": state,
            "message": message ?? "",
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onInitCompleted", arguments: payload)
        }
    }

    static let onStopCompleted: OnStopCompleted = {
        print("SDK stopped.")
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onStopCompleted", arguments: nil)
        }
    }

    static let onRegistrarState: OnRegistrarState = { state in
        print("Registrar state: \(state)")
        // 注册状态改变
        let payload: [String: Any] = [
            "state": state,
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onRegistrarState", arguments: payload)
        }
    }

    static let onIncomingCall: OnIncomingCall = { callParam in
        print("Incoming call with params: \(callParam)")
        var headers = [String: String]()
        // 使用 withUnsafePointer 获取 headers 指针
        withUnsafePointer(to: callParam.headers) { headersPtr in
            // 将 headers 指针转换为原始字节缓冲区
            let rawPtr = UnsafeRawPointer(headersPtr)
            // 绑定内存为 sip_header 类型
            let buffer = rawPtr.bindMemory(to: sip_header.self, capacity: Int(SDK_MAX_CUSTOM_HEADERS))
            // 遍历数组
            for i in 0 ..< SDK_MAX_CUSTOM_HEADERS {
                let header = buffer[Int(i)]
                if let name = header.key, let value = header.value {
                    headers[String(cString: name)] = String(cString: value)
                }
            }
        }

        let payload: [String: Any] = [
            "callType": callParam.call_type,
            "username": String(cString: callParam.username),
            "remoteIp": String(cString: callParam.remote_ip),
            "headers": headers,
            "callUuid": "\(callParam.call_uuid)", // 转成 String
            "transmitVideo": callParam.transmit_video,
            "transmitSound": callParam.transmit_sound,
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onIncomingCall", arguments: payload)
        }
    }

    static let onFindIncoming: OnFindIncoming = { type, findParam in
        let param = SIPFindIncomingParam(cParam: findParam)
        let payload: [String: Any] = [
            "currentType": type?.pointee ?? -1,
            "transportType": param.transportType,
            "transportName": param.transportName,
            "toDomain": param.toDomain,
            "toUsername": param.toUsername,
            "fromDomain": param.fromDomain,
            "fromUsername": param.fromUsername,
            "requestDomain": param.requestDomain,
            "requestUsername": param.requestUsername,
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onFindIncoming", arguments: payload)
        }
    }

    static let onDtmfInfo: OnDtmfInfo = { dtmfInfoParam in
        print("DTMF info: \(dtmfInfoParam)")
        // 注册状态改变
        var contentType: String?
        if let content_type = dtmfInfoParam.content_type {
            contentType = String(cString: content_type)
        }
        var content: String?
        if let c = dtmfInfoParam.content {
            content = String(cString: c)
        }
        let payload: [String: Any] = [
            "callUuid": "\(dtmfInfoParam.call_uuid)",
            "dtmfInfoType": dtmfInfoParam.dtmf_info_type,
            "contentType": contentType ?? "",
            "content": content ?? "",
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onDtmfInfo", arguments: payload)
        }
    }

    static let onMessage: OnMessage = { messageParam in
        print("Received message: \(messageParam)")
        var username: String?
        if let u = messageParam.username {
            username = String(cString: u)
        }
        var remoteIp: String?
        if let ip = messageParam.remote_ip {
            remoteIp = String(cString: ip)
        }
        var content: String?
        if let c = messageParam.content {
            content = String(cString: c)
        }
        let payload: [String: Any] = [
            "messageType": messageParam.message_type,
            "username": username ?? "",
            "remoteIp": remoteIp ?? "",
            "content": content ?? "",
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onMessage", arguments: payload)
        }
    }

    static let onMessageState: OnMessageState = { state, messageParam in
        print("Message state: \(state), \(messageParam)")
        var username: String?
        if let u = messageParam.username {
            username = String(cString: u)
        }
        var remoteIp: String?
        if let ip = messageParam.remote_ip {
            remoteIp = String(cString: ip)
        }
        var content: String?
        if let c = messageParam.content {
            content = String(cString: c)
        }
        let payload: [String: Any] = [
            "state": state,
            "message": [
                "messageType": messageParam.message_type,
                "username": username ?? "",
                "remoteIp": remoteIp ?? "",
                "content": content ?? "",
            ],
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onMessageState", arguments: payload)
        }
    }

    static let onCallState: OnCallState = { param in
        print("Call state changed. callUuid: \(param.call_uuid), State: \(param.state)")
        var lastStatusText: String?
        if let text = param.last_status_text {
            lastStatusText = String(cString: text)
        }
        // 呼叫状态改变
        NotificationCenter.default.post(
            name: .SIP_CALL_STATE_CHANGE,
            object: nil,
            userInfo: ["callUuid": param.call_uuid, "state": param.state, "lastStatus": param.last_status, "lastStatusText": lastStatusText ?? ""]
        )

        let payload: [String: Any] = [
            "callUuid": "\(param.call_uuid)",
            "state": param.state,
            "lastStatus": param.last_status,
            "lastStatusText": lastStatusText ?? "",
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onCallState", arguments: payload)
        }
    }

    static let onExpireWarning: ExpireWarningCallback = { expireTime, currentTime in
        let dateFormatter = ISO8601DateFormatter()
        let expireDate = Date(timeIntervalSince1970: TimeInterval(expireTime))
        let currentDate = Date(timeIntervalSince1970: TimeInterval(currentTime))

        let payload: [String: Any] = [
            "expireTime": dateFormatter.string(from: expireDate),
            "currentTime": dateFormatter.string(from: currentDate),
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onExpireWarning", arguments: payload)
        }
    }

    static let onActivityCheck: ActivityCheckCallback = {
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onActivityCheck", arguments: [])
        }
    }

    // MARK: - 文件传输（FT）回调

    static let onFTOffer: OnFTOffer = { ptr in
        guard let ptr = ptr else { return }
        let param = SIPFTOfferParam(cParam: ptr.pointee)
        let payload: [String: Any] = [
            "ftId": param.ftId,
            "accUuid": param.accUuid,
            "username": param.username,
            "remoteIp": param.remoteIp,
            "file": [
                "name": param.file.name,
                "size": param.file.size,
                "extra": param.file.extra,
            ],
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onFTOffer", arguments: payload)
        }
    }

    static let onFTRequest: OnFTRequest = { ptr in
        guard let ptr = ptr else { return }
        let info = SIPFTRequestInfo(cParam: ptr.pointee)
        let payload: [String: Any] = [
            "reqId": info.reqId,
            "accUuid": info.accUuid,
            "username": info.username,
            "remoteIp": info.remoteIp,
            "file": [
                "name": info.file.name,
                "size": info.file.size,
                "extra": info.file.extra,
            ],
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onFTRequest", arguments: payload)
        }
    }

    static let onFTRequestResult: OnFTRequestResult = { reqId, ok, reason in
        let reasonText = reason.map { String(cString: $0) } ?? ""
        let payload: [String: Any] = [
            "reqId": reqId,
            "ok": ok != 0,
            "reason": reasonText,
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onFTRequestResult", arguments: payload)
        }
    }

    static let onFTProgress: OnFTProgress = { ptr in
        guard let ptr = ptr else { return }
        let progress = SIPFTProgress(cParam: ptr.pointee)
        let payload: [String: Any] = [
            "ftId": progress.ftId,
            "state": progress.state,
            "bytesTotal": progress.bytesTotal,
            "bytesDone": progress.bytesDone,
            "percent": progress.percent,
            "activeSessions": progress.activeSessions,
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onFTProgress", arguments: payload)
        }
    }

    static let onFTComplete: OnFTComplete = { ptr in
        guard let ptr = ptr else { return }
        let param = SIPFTCompleteParam(cParam: ptr.pointee)
        let payload: [String: Any] = [
            "ftId": param.ftId,
            "role": param.role,
            "error": param.error,
            "bytesTransferred": param.bytesTransferred,
            "elapsedMs": param.elapsedMs,
            "fileName": param.fileName,
            "savePath": param.savePath,
        ]
        DispatchQueue.main.async {
            SipSdkFlutterPlugin.channel?.invokeMethod("onFTComplete", arguments: payload)
        }
    }

    /// 配置 FT（构建 SIPFTCallbacks 并传给 SIPHandle.setFTConfig）
    static func setupFTConfig(config: SIPFTConfig) -> Int32 {
        let callbacks = SIPFTCallbacks(
            onFTOffer: onFTOffer,
            onFTRequest: onFTRequest,
            onFTRequestResult: onFTRequestResult,
            onFTProgress: onFTProgress,
            onFTComplete: onFTComplete
        )
        return SIPHandle.setFTConfig(config: config, callbacks: callbacks)
    }
}
